File: src/bot.ts
import dotenv from 'dotenv';
import TelegramBot from 'node-telegram-bot-api';
import axios from 'axios';

dotenv.config();

const BOT_TOKEN = process.env.BOT_TOKEN!;
const API_BASE = process.env.API_BASE_URL!;

if (!BOT_TOKEN || !API_BASE) {
  console.error('Error: BOT_TOKEN and API_BASE_URL must be set in .env');
  process.exit(1);
}

const bot = new TelegramBot(BOT_TOKEN, { polling: true });

// /search command
bot.onText(/\/search (.+)/, async (msg, match) => {
  const chatId = msg.chat.id;
  const query = match![1].trim();

  try {
    const res = await axios.get(`${API_BASE}/search`, {
      params: { key: query }
    });
    const results: Array<{ title: string; id: string }> = res.data.data || res.data;
    if (!results.length) {
      return bot.sendMessage(chatId, `No results found for "${query}".`);
    }

    // List top 5
    const top5 = results.slice(0, 5);
    const list = top5
      .map((anime, i) => `${i + 1}. ${anime.title} (slug: ${anime.id})`)
      .join('\n');

    await bot.sendMessage(
      chatId,
      `Top results for "${query}":\n\n${list}`
    );
  } catch (err) {
    console.error(err);
    bot.sendMessage(chatId, 'Error searching anime. Please try again later.');
  }
});

// /get command
bot.onText(/\/get (\S+) (\d+)/, async (msg, match) => {
  const chatId = msg.chat.id;
  const animeSlug = match![1];
  const epNum = parseInt(match![2], 10);

  try {
    // Fetch episodes
    const epsRes = await axios.get(`${API_BASE}/episodes/${animeSlug}`);
    const episodes: Array<{ id: string; title: string }> = epsRes.data;
    const ep = episodes.find(e => {
      const m = e.title.match(/Episode\s*(\d+)/i);
      return m && parseInt(m[1], 10) === epNum;
    });
    if (!ep) {
      return bot.sendMessage(chatId, `Episode ${epNum} not found for ${animeSlug}.`);
    }

    // Fetch servers
    const srvRes = await axios.get(`${API_BASE}/servers`, { params: { id: ep.id } });
    const hd2 = srvRes.data.sub.find((s: any) => s.name === 'HD-2');
    if (!hd2) {
      return bot.sendMessage(chatId, 'HD-2 server not available for this episode.');
    }

    // Fetch stream & subtitles
    const strRes = await axios.get(`${API_BASE}/stream`, {
      params: { id: ep.id, server: hd2.name, type: 'sub' }
    });
    const { streamingLink, subtitles } = strRes.data;

    await bot.sendMessage(
      chatId,
      `📥 *Download Link*:\n${streamingLink}`,
      { parse_mode: 'Markdown' }
    );

    if (subtitles?.length) {
      const subMsg = subtitles
        .map((s: any) => `💬 *Subtitle (${s.lang})*:\n${s.src}`)
        .join('\n\n');
      await bot.sendMessage(chatId, subMsg, { parse_mode: 'Markdown' });
    } else {
      await bot.sendMessage(chatId, 'No English subtitles found.', { parse_mode: 'Markdown' });
    }
  } catch (err) {
    console.error(err);
    bot.sendMessage(chatId, 'An error occurred. Please try again later.');
  }
});

console.log('Bot is up and running.');

File: src/bot.py
import os
import logging
import requests
from dotenv import load_dotenv
from telegram import Update
from telegram.constants import ParseMode
from telegram.ext import ApplicationBuilder, CommandHandler, ContextTypes

# Load env vars
load_dotenv()
BOT_TOKEN    = os.getenv('BOT_TOKEN')
API_BASE_URL = os.getenv('API_BASE_URL')  # e.g. https://hiapitest.onrender.com/api/v1

if not BOT_TOKEN or not API_BASE_URL:
    print("Error: BOT_TOKEN and API_BASE_URL must be set in .env")
    exit(1)

# Configure logging
logging.basicConfig(
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
    level=logging.INFO
)
logger = logging.getLogger(__name__)

# === /search handler (unchanged) ===
async def search(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    if not context.args:
        return await update.message.reply_text('Usage: /search <anime name>')

    query = ' '.join(context.args)
    # note: this API expects `keyword`
    resp = requests.get(f"{API_BASE_URL}/search", params={'keyword': query})
    raw = {}
    try:
        raw = resp.json()
    except ValueError:
        pass

    if resp.status_code != 200 or not raw.get('success', False):
        logger.error("Search failed (%s): %s", resp.status_code, raw)
        return await update.message.reply_text(f'No results for "{query}".')

    results = raw.get('data', {}).get('response', [])
    if not isinstance(results, list) or not results:
        return await update.message.reply_text(f'No results for "{query}".')

    top5 = results[:5]
    lines = [
        f"{i+1}. {item.get('title','–')} (slug: {item.get('id','–')})"
        for i, item in enumerate(top5)
    ]
    await update.message.reply_text(
        f"Top results for '{query}':\n\n" + "\n".join(lines)
    )

# === /get handler (with slug‐cleaning) ===
async def get_episode(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    if len(context.args) < 2:
        return await update.message.reply_text(
            'Usage: /get <animeSlug> <episodeNumber>'
        )

    # Strip off any query‐string from the slug
    raw_slug = context.args[0]
    anime_slug = raw_slug.split('?', 1)[0]
    try:
        ep_num = int(context.args[1])
    except ValueError:
        return await update.message.reply_text('Episode number must be an integer.')

    try:
        # 1) Fetch episodes list
        eps_res = requests.get(f"{API_BASE_URL}/episodes/{anime_slug}")
        eps_res.raise_for_status()
        episodes = eps_res.json()

        # Find the matching episode object
        ep_item = next(
            (e for e in episodes if f"Episode {ep_num}" in e.get('title', "")),
            None
        )
        if not ep_item:
            return await update.message.reply_text(
                f'Episode {ep_num} not found for "{anime_slug}".'
            )

        # 2) Fetch available servers
        srv_res = requests.get(
            f"{API_BASE_URL}/servers",
            params={'id': ep_item['id']}
        )
        srv_res.raise_for_status()
        hd2 = next(
            (s for s in srv_res.json().get('sub', []) if s.get('name') == 'HD-2'),
            None
        )
        if not hd2:
            return await update.message.reply_text(
                'HD-2 server not available for this episode.'
            )

        # 3) Fetch stream link + subtitles
        str_res = requests.get(
            f"{API_BASE_URL}/stream",
            params={
                'id': ep_item['id'],
                'server': hd2['name'],
                'type': 'sub'
            }
        )
        str_res.raise_for_status()
        data = str_res.json()

        # Send download link
        await update.message.reply_text(
            f"📥 *Download Link*:\n{data['streamingLink']}",
            parse_mode=ParseMode.MARKDOWN
        )

        # Send any English subtitles
        subs = data.get('subtitles', [])
        if subs:
            for sub in subs:
                await update.message.reply_text(
                    f"💬 *Subtitle ({sub['lang']})*:\n{sub['src']}",
                    parse_mode=ParseMode.MARKDOWN
                )
        else:
            await update.message.reply_text('No English subtitles found.')

    except Exception as e:
        logger.error("Get error: %s", e)
        await update.message.reply_text(
            'An error occurred while fetching the episode. Please try again later.'
        )

# === Main ===
if __name__ == '__main__':
    app = ApplicationBuilder().token(BOT_TOKEN).build()
    app.add_handler(CommandHandler('search', search))
    app.add_handler(CommandHandler('get', get_episode))
    logger.info('Bot is starting...')
    app.run_polling()
